libvast/src/system/query_supervisor.cpp

```diff
@@ -54,31 +54,35 @@ query_supervisor_actor::behavior_type query_supervisor(
       VAST_DEBUG(self, "got a new query for", qm.size(),
                  "partitions:", get_ids(qm));
       VAST_ASSERT(!qm.empty());
-      VAST_ASSERT(self->state.open_requests.empty());
+      VAST_ASSERT(self->state.open_requests == 0);
       for (auto& [id, partition] : qm) {
-        self->state.open_requests.emplace(id, 1 /*qm.size()*/);
+        ++self->state.open_requests;
         // TODO: Add a proper configurable timeout.
-        // TODO: Handle the error case for the `then()` handler.
         self
           ->request(partition, caf::infinite, expr,
                     caf::actor_cast<partition_client_actor>(client))
-          .then([=, id = id](atom::done) {
-            auto& num_evaluators = self->state.open_requests[id];
-            if (--num_evaluators == 0) {
-              VAST_DEBUG(self, "collected all results for partition", id);
-              self->state.open_requests.erase(id);
-              // Ask master for more work after receiving the last sub
-              // result.
-              // TODO: We should schedule a new partition as soon as one has
-              // finished, otherwise each batch will be as slow as the slowest
-              // contained partition.
-              if (self->state.open_requests.empty()) {
-                VAST_DEBUG(self, "collected all results for all partitions");
+          .then(
+            [=, id = id](atom::done) {
+              if (--self->state.open_requests == 0) {
+                VAST_DEBUG(self, "collected all results for partition", id);
```

Comment:
I don't think the capture of `id` makes sense here. If my understanding is correct the debug message will always print the id of the partition that was requested last. I'd say to change the log message to `VAST_DEBUG(self, "collected all result for the current batch of partitions");`.

Reply:
Maybe I'm missing something, but aren't we copying the the `id` of the current iteration each time?

Reply:
Yes, each instance of the lambda gets an id, but only the one from the last partition is printed in the log. You could also move the log statement before the condition to change that.

Reply:
Ah, right, updated.

Reply:
`... for all partitions` is misleading, because is suggests that the index is done with that query, but this message will be shown each time a subset of the candidates is done, e.g. after "request more hits".